go/teams/service_helper.go

```diff
@@ -448,65 +448,103 @@ type AddMembersRes struct {
 
 // AddMembers adds a bunch of people to a team. Assertions can contain usernames or social assertions.
 // Adds them all in a transaction so it's all or nothing.
-// On success, returns a list where len(res)=len(assertions) and in corresponding order.
-func AddMembers(ctx context.Context, g *libkb.GlobalContext, teamID keybase1.TeamID, users []keybase1.UserRolePair) (res []AddMembersRes, err error) {
+// If the first transaction fails due to TeamContactSettingsBlock error, it
+// will remove restricted users returned by the error, and retry once.
+// On success, returns a list where len(added) + len(noAdded) = len(assertions) and in
+// corresponding order, with restricted users having an empty AddMembersRes.
+func AddMembers(ctx context.Context, g *libkb.GlobalContext, teamID keybase1.TeamID, users []keybase1.UserRolePair) (added []AddMembersRes, notAdded []keybase1.User, err error) {
+	mctx := libkb.NewMetaContext(ctx, g)
 	tracer := g.CTimeTracer(ctx, "team.AddMembers", true)
 	defer tracer.Finish()
 
-	err = RetryIfPossible(ctx, g, func(ctx context.Context, _ int) error {
-		res = make([]AddMembersRes, len(users))
-		team, err := GetForTeamManagementByTeamID(ctx, g, teamID, true /*needAdmin*/)
-		if err != nil {
-			return err
-		}
+	addNonRestrictedMembersFunc := func(restrictedUsers map[keybase1.UID]bool) func(ctx context.Context, _ int) error {
+		return func(ctx context.Context, _ int) error {
+			added = []AddMembersRes{}
+			notAdded = []keybase1.User{}
 
-		tx := CreateAddMemberTx(team)
-		type sweepEntry struct {
-			Assertion string
-			UV        keybase1.UserVersion
-		}
-		var sweep []sweepEntry
-		for i, user := range users {
-			username, uv, invite, err := tx.AddMemberByAssertionOrEmail(ctx, user.AssertionOrEmail, user.Role, user.BotSettings)
+			team, err := GetForTeamManagementByTeamID(ctx, g, teamID, true /*needAdmin*/)
 			if err != nil {
-				if _, ok := err.(AttemptedInviteSocialOwnerError); ok {
-					return err
-				}
-				return NewAddMembersError(user.AssertionOrEmail, err)
-			}
-			var normalizedUsername libkb.NormalizedUsername
-			if !username.IsNil() {
-				normalizedUsername = username
+				return err
 			}
-			res[i] = AddMembersRes{
-				Invite:   invite,
-				Username: normalizedUsername,
+
+			tx := CreateAddMemberTx(team)
+			type sweepEntry struct {
+				Assertion string
+				UV        keybase1.UserVersion
 			}
-			if !uv.IsNil() {
-				sweep = append(sweep, sweepEntry{
-					Assertion: user.AssertionOrEmail,
-					UV:        uv,
+			var sweep []sweepEntry
+			for _, user := range users {
+				upak, single, doInvite, err := tx.ResolveUPKV2FromAssertionOrEmail(mctx, user.AssertionOrEmail)
```

Comment:
As discussed in kb chat - it might be better for `upak, single, doInvite` to live in a struct that's returned from `ResolveUPKV2FromAssertionOrEmail` and passed to `AddOrInviteMemberByUPKV2`.